File: TS/records.ts
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";
import {
  createQuickbaseClient,
  handleApiError,
  truncateIfNeeded,
  formatRecordsAsMarkdown,
  QuickbaseConfig,
  QuickbaseQueryResponse,
  QuickbaseUpsertResponse,
} from "../services/quickbase.js";

export function registerRecordTools(server: McpServer, config: QuickbaseConfig): void {
  const client = createQuickbaseClient(config);

  // ─── Query Records ──────────────────────────────────────────────
  server.registerTool(
    "quickbase_query_records",
    {
      title: "Query Records",
      description: `Query records from a Quickbase table with optional filtering, sorting, and pagination.

Uses Quickbase's query syntax where conditions follow the format: {fieldId.operator.value}
Common operators: EX (equals), CT (contains), GT (greater than), LT (less than), GTE, LTE, BEF (date before), AF (date after)
Combine with AND/OR: {6.EX.'Active'} AND {7.CT.'John'}

Args:
  - tableId (string): The Quickbase table ID (e.g. "bsf8tqhe6")
  - where (string, optional): Query filter string. Leave blank to get all records.
  - select (number[], optional): Field IDs to return. Leave blank for all fields.
  - sortBy (object[], optional): Array of {fieldId, order} for sorting. order: "ASC" | "DESC"
  - skip (number, optional): Number of records to skip for pagination (default: 0)
  - top (number, optional): Max records to return (default: 100, max: 10000)
  - response_format (string): "markdown" or "json" (default: "markdown")

Returns:
  Matching records with field labels and metadata (total count, pagination info).

Examples:
  - All records: leave where empty, set top: 50
  - Filter active users: where = "{8.EX.'Active'}"
  - Date range: where = "{10.AF.'2024-01-01'} AND {10.BEF.'2024-12-31'}"`,
      inputSchema: z.object({
        tableId: z.string().min(1).describe("Quickbase table ID (e.g. 'bsf8tqhe6')"),
        where: z.string().optional().describe("Query filter, e.g. '{6.EX.\"Active\"}'"),
        select: z.array(z.number().int().positive()).optional().describe("Field IDs to return"),
        sortBy: z.array(z.object({
          fieldId: z.number().int().positive(),
          order: z.enum(["ASC", "DESC"]).default("ASC"),
        })).optional().describe("Sort order"),
        skip: z.number().int().min(0).default(0).describe("Records to skip for pagination"),
        top: z.number().int().min(1).max(10000).default(100).describe("Max records to return"),
        response_format: z.enum(["markdown", "json"]).default("markdown"),
      }).strict(),
      annotations: { readOnlyHint: true, destructiveHint: false, idempotentHint: true, openWorldHint: true },
    },
    async ({ tableId, where, select, sortBy, skip, top, response_format }) => {
      try {
        const body: Record<string, unknown> = { from: tableId, options: { skip, top } };
        if (where) body.where = where;
        if (select?.length) body.select = select;
        if (sortBy?.length) body.sortBy = sortBy.map((s) => ({ fieldId: s.fieldId, order: s.order }));

        const { data } = await client.post<QuickbaseQueryResponse>("/records/query", body);
        const records = data.data ?? [];
        const fields = data.fields ?? [];
        const meta = data.metadata;

        if (response_format === "json") {
          const output = { totalRecords: meta.totalRecords, returned: meta.numRecords, skip: meta.skip, records, fields };
          return { content: [{ type: "text", text: truncateIfNeeded(JSON.stringify(output, null, 2)) }] };
        }

        const table = formatRecordsAsMarkdown(records, fields);
        const summary = `**Table:** \`${tableId}\` | **Total:** ${meta.totalRecords} records | **Showing:** ${meta.numRecords} (skip: ${meta.skip})\n\n${table}`;
        return { content: [{ type: "text", text: truncateIfNeeded(summary) }] };
      } catch (err) {
        return { content: [{ type: "text", text: `Error: ${handleApiError(err)}` }], isError: true };
      }
    }
  );

  // ─── Get Record by ID ────────────────────────────────────────────
  server.registerTool(
    "quickbase_get_record",
    {
      title: "Get Record by ID",
      description: `Retrieve a single record by its Record ID from a Quickbase table.

Args:
  - tableId (string): The Quickbase table ID
  - recordId (number): The record's unique ID (field 3 is always the Record ID)
  - select (number[], optional): Specific field IDs to return

Returns:
  The record's field values with labels. Returns error if the record does not exist.`,
      inputSchema: z.object({
        tableId: z.string().min(1).describe("Quickbase table ID"),
        recordId: z.number().int().positive().describe("Record ID number"),
        select: z.array(z.number().int().positive()).optional().describe("Field IDs to return"),
      }).strict(),
      annotations: { readOnlyHint: true, destructiveHint: false, idempotentHint: true, openWorldHint: true },
    },
    async ({ tableId, recordId, select }) => {
      try {
        const body: Record<string, unknown> = {
          from: tableId,
          where: `{3.EX.${recordId}}`,
        };
        if (select?.length) body.select = select;

        const { data } = await client.post<QuickbaseQueryResponse>("/records/query", body);
        const records = data.data ?? [];

        if (records.length === 0) {
          return { content: [{ type: "text", text: `No record found with ID ${recordId} in table ${tableId}.` }] };
        }

        const fields = data.fields ?? [];
        const record = records[0];
        const lines = fields.map((f) => {
          const val = record[String(f.id)]?.value;
          return `**${f.label}** (field ${f.id}): ${val === null || val === undefined ? "_empty_" : typeof val === "object" ? JSON.stringify(val) : String(val)}`;
        });

        return { content: [{ type: "text", text: `**Record ${recordId} in \`${tableId}\`**\n\n${lines.join("\n")}` }] };
      } catch (err) {
        return { content: [{ type: "text", text: `Error: ${handleApiError(err)}` }], isError: true };
      }
    }
  );

  // ─── Upsert Records ──────────────────────────────────────────────
  server.registerTool(
    "quickbase_upsert_records",
    {
      title: "Create or Update Records",
      description: `Insert new records or update existing records in a Quickbase table (upsert).

To CREATE a new record: omit the Record ID field (field 3) from the data.
To UPDATE an existing record: include field 3 with the record's ID.
To use a custom merge field: specify mergeFieldId to match on that field instead.

Args:
  - tableId (string): Target Quickbase table ID
  - data (object[]): Array of records. Each record is { fieldId: value } pairs.
      Example: [{ "6": "Active", "7": "John Smith", "8": 42 }]
  - mergeFieldId (number, optional): Field ID to use as upsert key (instead of Record ID)
  - fieldsToReturn (number[], optional): Field IDs to include in the response

Returns:
  Summary of created, updated, and unchanged record IDs.

Examples:
  - Create: data = [{ "6": "New Value", "7": "Another Value" }]
  - Update record 5: data = [{ "3": 5, "6": "Updated Value" }]`,
      inputSchema: z.object({
        tableId: z.string().min(1).describe("Target table ID"),
        data: z.array(z.record(z.string(), z.unknown())).min(1).describe("Array of records as {fieldId: value} objects"),
        mergeFieldId: z.number().int().positive().optional().describe("Field ID to match on for upsert (default: Record ID)"),
        fieldsToReturn: z.array(z.number().int().positive()).optional().describe("Field IDs to return in response"),
      }).strict(),
      annotations: { readOnlyHint: false, destructiveHint: false, idempotentHint: false, openWorldHint: true },
    },
    async ({ tableId, data, mergeFieldId, fieldsToReturn }) => {
      try {
        // Transform data: wrap values in { value: ... } as required by Quickbase API
        const transformedData = data.map((record) => {
          const transformed: Record<string, { value: unknown }> = {};
          for (const [fieldId, value] of Object.entries(record)) {
            transformed[fieldId] = { value };
          }
          return transformed;
        });

        const body: Record<string, unknown> = { to: tableId, data: transformedData };
        if (mergeFieldId) body.mergeFieldId = mergeFieldId;
        if (fieldsToReturn?.length) body.fieldsToReturn = fieldsToReturn;

        const { data: result } = await client.post<QuickbaseUpsertResponse>("/records", body);
        const meta = result.metadata;

        const summary = [
          `**Upsert complete for table \`${tableId}\`**`,
          `- ✅ Created: ${meta.createdRecordIds.length} records (IDs: ${meta.createdRecordIds.join(", ") || "none"})`,
          `- 🔄 Updated: ${meta.updatedRecordIds.length} records (IDs: ${meta.updatedRecordIds.join(", ") || "none"})`,
          `- ⏭️  Unchanged: ${meta.unchangedRecordIds.length} records`,
          `- 📊 Total processed: ${meta.totalNumberOfRecordsProcessed}`,
        ].join("\n");

        return { content: [{ type: "text", text: summary }] };
      } catch (err) {
        return { content: [{ type: "text", text: `Error: ${handleApiError(err)}` }], isError: true };
      }
    }
  );

  // ─── Delete Records ──────────────────────────────────────────────
  server.registerTool(
    "quickbase_delete_records",
    {
      title: "Delete Records",
      description: `Delete records from a Quickbase table matching a query condition. This operation is irreversible.

Uses Quickbase query syntax for the where clause.
CAUTION: If you want to delete all records, use where = "{'3'.GT.'0'}" (all record IDs greater than 0).

Args:
  - tableId (string): Table ID containing the records to delete
  - where (string): Required filter query. Records matching this will be deleted.
      Example: "{3.EX.42}" deletes record with ID 42
      Example: "{6.EX.'Inactive'}" deletes all records where field 6 equals "Inactive"

Returns:
  Number of records deleted.`,
      inputSchema: z.object({
        tableId: z.string().min(1).describe("Table ID"),
        where: z.string().min(1).describe("Query to select records for deletion, e.g. '{3.EX.42}'"),
      }).strict(),
      annotations: { readOnlyHint: false, destructiveHint: true, idempotentHint: false, openWorldHint: true },
    },
    async ({ tableId, where }) => {
      try {
        const { data } = await client.delete<{ numberDeleted: number }>("/records", {
          data: { from: tableId, where },
        });
        return { content: [{ type: "text", text: `✅ Deleted **${data.numberDeleted}** record(s) from table \`${tableId}\`.` }] };
      } catch (err) {
        return { content: [{ type: "text", text: `Error: ${handleApiError(err)}` }], isError: true };
      }
    }
  );
}

File: py/tools/records.py
"""
Quickbase record tools.

Tools:
  - qb_query_records  (P0) — query records with optional filters
  - qb_create_record  (P0) — insert a new record
  - qb_update_record  (P0) — update fields on an existing record
  - qb_delete_record  (P0) — delete a record by record ID
"""

import json
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field, ConfigDict, field_validator
from mcp.server.fastmcp import FastMCP

from client import qb_request, handle_api_error


def _build_field_value(value: Any) -> Dict[str, Any]:
    """Wrap a plain Python value in Quickbase's {"value": ...} envelope."""
    return {"value": value}


def _format_records(records: List[Dict]) -> List[Dict]:
    """
    Flatten Quickbase's field-envelope format to plain dicts.

    Input:  [{"3": {"value": "Alice"}, "6": {"value": 42}}, ...]
    Output: [{"3": "Alice", "6": 42}, ...]
    """
    return [
        {str(fid): fdata.get("value") for fid, fdata in record.items()}
        for record in records
    ]


def register_record_tools(mcp: FastMCP) -> None:
    """Register all record-level tools onto the FastMCP server instance."""

    # ── qb_query_records (P0) ────────────────────────────────────────────────

    class QueryRecordsInput(BaseModel):
        """Input for querying Quickbase records."""

        model_config = ConfigDict(str_strip_whitespace=True, extra="forbid")

        table_id: str = Field(
            ...,
            description="The Quickbase table ID to query (e.g. 'bqxyz.1').",
            min_length=1,
        )
        where: Optional[str] = Field(
            default=None,
            description=(
                "Quickbase formula-style filter string, e.g. \"{6}.EX.'open'\". "
                "Omit to return all records (subject to limit)."
            ),
        )
        select: Optional[List[int]] = Field(
            default=None,
            description=(
                "List of field IDs to include in results. "
                "Omit to return all fields. Use qb_get_table_schema to discover field IDs."
            ),
        )
        sort_by: Optional[List[Dict[str, Any]]] = Field(
            default=None,
            description=(
                "Sort specification, e.g. [{\"fieldId\": 6, \"order\": \"ASC\"}]."
            ),
        )
        limit: int = Field(
            default=100,
            description="Maximum number of records to return (1–1000).",
            ge=1,
            le=1000,
        )
        skip: int = Field(
            default=0,
            description="Number of records to skip for pagination.",
            ge=0,
        )

    @mcp.tool(
        name="qb_query_records",
        annotations={
            "title": "Query Quickbase Records",
            "readOnlyHint": True,
            "destructiveHint": False,
            "idempotentHint": True,
            "openWorldHint": True,
        },
    )
    async def qb_query_records(params: QueryRecordsInput) -> str:
        """
        Query records in a Quickbase table with optional filters, field selection, and sorting.

        Uses the Quickbase Records Query API (POST /v1/records/query). Filters use
        Quickbase's formula syntax: {fieldId}.OPERATOR.'value'.

        Args:
            params (QueryRecordsInput):
                - table_id (str): Table to query.
                - where (str, optional): Filter formula (e.g. "{6}.EX.'open'").
                - select (List[int], optional): Field IDs to return.
                - sort_by (List[dict], optional): Sort order.
                - limit (int): Max records to return (default 100, max 1000).
                - skip (int): Records to skip for pagination (default 0).

        Returns:
            str: JSON object:
                {
                    "fields": [{"id": int, "label": str, "type": str}, ...],
                    "records": [{"<fieldId>": <value>, ...}, ...],
                    "metadata": {
                        "totalRecords": int,
                        "numRecords": int,
                        "skip": int,
                        "hasMore": bool
                    }
                }

                Error: "<message>" on failure.

        Examples:
            - Use when: "Get all open tickets from the Issues table"
              → params with where="{status_field_id}.EX.'open'"
            - Use when: "Show me the last 10 records sorted by date"
              → params with limit=10, sort_by=[{"fieldId": <date_field_id>, "order": "DESC"}]
        """
        try:
            body: Dict[str, Any] = {
                "from": params.table_id,
                "options": {"skip": params.skip, "top": params.limit},
            }
            if params.where:
                body["where"] = params.where
            if params.select:
                body["select"] = params.select
            if params.sort_by:
                body["sortBy"] = params.sort_by

            data = await qb_request("POST", "/records/query", json=body)

            fields = data.get("fields", [])
            raw_records = data.get("data", [])
            metadata = data.get("metadata", {})

            flat_records = _format_records(raw_records)
            total = metadata.get("totalRecords", len(flat_records))
            num = metadata.get("numRecords", len(flat_records))

            result = {
                "fields": [
                    {"id": f.get("id"), "label": f.get("label"), "type": f.get("type")}
                    for f in fields
                ],
                "records": flat_records,
                "metadata": {
                    "totalRecords": total,
                    "numRecords": num,
                    "skip": params.skip,
                    "hasMore": (params.skip + num) < total,
                },
            }
            return json.dumps(result, indent=2)
        except Exception as e:
            return handle_api_error(e)

    # ── qb_create_record (P0) ────────────────────────────────────────────────

    class CreateRecordInput(BaseModel):
        """Input for creating a new Quickbase record."""

        model_config = ConfigDict(extra="forbid")

        table_id: str = Field(
            ...,
            description="The table ID to insert the record into.",
            min_length=1,
        )
        fields: Dict[str, Any] = Field(
            ...,
            description=(
                "Map of field ID (as string) to value, e.g. {\"6\": \"open\", \"7\": \"Bug report\"}. "
                "Use qb_get_table_schema to find field IDs and types."
            ),
        )

        @field_validator("fields")
        @classmethod
        def fields_not_empty(cls, v: Dict) -> Dict:
            if not v:
                raise ValueError("fields must contain at least one field/value pair.")
            return v

    @mcp.tool(
        name="qb_create_record",
        annotations={
            "title": "Create a Quickbase Record",
            "readOnlyHint": False,
            "destructiveHint": False,
            "idempotentHint": False,
            "openWorldHint": True,
        },
    )
    async def qb_create_record(params: CreateRecordInput) -> str:
        """
        Insert a new record into a Quickbase table.

        Args:
            params (CreateRecordInput):
                - table_id (str): Table to insert into.
                - fields (dict): Field ID → value map (field IDs as strings).

        Returns:
            str: JSON object on success:
                {
                    "recordId": int,      # The new record's ID
                    "createdRecordIds": [int],
                    "unchangedRecordIds": [],
                    "updatedRecordIds": []
                }

                Error: "<message>" on failure.

        Examples:
            - Use when: "Create a new ticket with status 'open' and title 'Login bug'"
              → params with fields={"6": "open", "7": "Login bug"}
        """
        try:
            # Build Quickbase field envelope format
            data_fields = {
                int(fid): _build_field_value(val)
                for fid, val in params.fields.items()
            }
            body = {"to": params.table_id, "data": [data_fields]}
            data = await qb_request("POST", "/records", json=body)

            created_ids = data.get("metadata", {}).get("createdRecordIds", [])
            result = {
                "recordId": created_ids[0] if created_ids else None,
                "createdRecordIds": created_ids,
                "unchangedRecordIds": data.get("metadata", {}).get("unchangedRecordIds", []),
                "updatedRecordIds": data.get("metadata", {}).get("updatedRecordIds", []),
            }
            return json.dumps(result, indent=2)
        except Exception as e:
            return handle_api_error(e)

    # ── qb_update_record (P0) ────────────────────────────────────────────────

    class UpdateRecordInput(BaseModel):
        """Input for updating an existing Quickbase record."""

        model_config = ConfigDict(extra="forbid")

        table_id: str = Field(..., description="The table ID containing the record.", min_length=1)
        record_id: int = Field(
            ...,
            description="The numeric record ID to update (field 3 in Quickbase).",
            ge=1,
        )
        fields: Dict[str, Any] = Field(
            ...,
            description=(
                "Map of field ID (as string) to new value. "
                "Only include fields you want to change."
            ),
        )

        @field_validator("fields")
        @classmethod
        def fields_not_empty(cls, v: Dict) -> Dict:
            if not v:
                raise ValueError("fields must contain at least one field/value pair.")
            return v

    @mcp.tool(
        name="qb_update_record",
        annotations={
            "title": "Update a Quickbase Record",
            "readOnlyHint": False,
            "destructiveHint": False,
            "idempotentHint": True,
            "openWorldHint": True,
        },
    )
    async def qb_update_record(params: UpdateRecordInput) -> str:
        """
        Update one or more fields on an existing Quickbase record.

        Args:
            params (UpdateRecordInput):
                - table_id (str): Table containing the record.
                - record_id (int): Record ID to update (the value of field 3).
                - fields (dict): Field ID → new value map.

        Returns:
            str: JSON object on success:
                {
                    "updatedRecordIds": [int],
                    "unchangedRecordIds": [int],
                    "createdRecordIds": []
                }

                Error: "<message>" on failure.

        Examples:
            - Use when: "Mark ticket #42 as resolved"
              → params with record_id=42, fields={"6": "resolved"}
        """
        try:
            data_fields = {int(fid): _build_field_value(val) for fid, val in params.fields.items()}
            # Quickbase requires record ID in field 3
            data_fields[3] = _build_field_value(params.record_id)
            body = {"to": params.table_id, "data": [data_fields]}
            data = await qb_request("POST", "/records", json=body)

            meta = data.get("metadata", {})
            result = {
                "updatedRecordIds": meta.get("updatedRecordIds", []),
                "unchangedRecordIds": meta.get("unchangedRecordIds", []),
                "createdRecordIds": meta.get("createdRecordIds", []),
            }
            return json.dumps(result, indent=2)
        except Exception as e:
            return handle_api_error(e)

    # ── qb_delete_record (P0) ────────────────────────────────────────────────

    class DeleteRecordInput(BaseModel):
        """Input for deleting a Quickbase record."""

        model_config = ConfigDict(extra="forbid")

        table_id: str = Field(..., description="The table ID containing the record.", min_length=1)
        record_ids: List[int] = Field(
            ...,
            description="One or more record IDs to delete.",
            min_length=1,
            max_length=100,
        )

    @mcp.tool(
        name="qb_delete_record",
        annotations={
            "title": "Delete Quickbase Record(s)",
            "readOnlyHint": False,
            "destructiveHint": True,
            "idempotentHint": False,
            "openWorldHint": True,
        },
    )
    async def qb_delete_record(params: DeleteRecordInput) -> str:
        """
        Delete one or more records from a Quickbase table by record ID.

        This operation is destructive and cannot be undone. Always confirm the
        record IDs before calling this tool.

        Args:
            params (DeleteRecordInput):
                - table_id (str): Table containing the records.
                - record_ids (List[int]): Record IDs to delete (up to 100 at a time).

        Returns:
            str: JSON object on success:
                {
                    "numberDeleted": int
                }

                Error: "<message>" on failure.

        Examples:
            - Use when: "Delete ticket #99 from the Issues table"
              → params with table_id="bqxyz.1", record_ids=[99]
        """
        try:
            body = {
                "from": params.table_id,
                "where": "{3}.EX." + ".OR.{3}.EX.".join(str(rid) for rid in params.record_ids),
            }
            data = await qb_request("DELETE", "/records", json=body)
            return json.dumps({"numberDeleted": data.get("numberDeleted", 0)}, indent=2)
        except Exception as e:
            return handle_api_error(e)
